feat(graphql): apply conjunction of filters

andrewazores opened the pull request:
Related to #820
Related to #846

jan-law commented on src/main/java/io/cryostat/net/web/http/api/beta/graph/EnvironmentNodesFetcher.java:
```diff
@@ -68,56 +66,24 @@
-                throw new NoSuchElementException(String.format("Node of type %s", nodeType));
-            }
-            return typedNodes;
+        if (filter.contains(FilterInput.Key.NODE_TYPE)) {
+            String nodeType = filter.get(FilterInput.Key.NODE_TYPE);
+            nodes = recurse(root, n -> n.getNodeType().getKind().equalsIgnoreCase(nodeType));
```
> The initial implementations of input objects will only support simple string-equality matches, so ex. { name: "myNode", kind: "JVM" } is an input object that logically means something like node.name.equals("myNode") && node.kind.equals("JVM")

Could you explain how this filters nodes with both `nodeName` `AND` `nodeType`? Wouldn't the filtered `nodes` from line 79 get overwritten here?

andrewazores replied:
Ah you're correct, good catch. The idea was to perform the query for all nodes first and then sequentially apply each filter, which has the same result as logically ANDing the filter criteria and applying that, but that's not what is actually implemented here. I'll fix it.

hareetd commented on src/main/java/io/cryostat/net/web/http/api/beta/graph/EnvironmentNodesFetcher.java:
```diff
@@ -38,68 +38,75 @@
-            EnvironmentNode node, Function<EnvironmentNode, Boolean> matcher) {
-        List<EnvironmentNode> result = new ArrayList<>();
+    Set<EnvironmentNode> flattenEnvNodes(EnvironmentNode root) {
+        return new HashSet<>(recurse(root, e -> e instanceof EnvironmentNode));
```
Is the switch to `HashSet` from `ArrayList` due to performance reasons? I guess in this case, the order doesn't matter and nodes are unique so `HashSet` makes more sense?

andrewazores replied:
Right, order doesn't matter/is undefined for this query. It isn't so much a performance reason as a cheap way to de-duplicate the collection. There shouldn't really be duplicates anyway if the tree is truly a tree and not a general graph (with potential cycles), but this is cheap insurance against that possibility.

hareetd commented:
While running `smoketest`, when I do the following query:

`query {
    environmentNodes(filter: { name: "JDP" }) {
        name
        nodeType
        descendantTargets {
            name
            nodeType
        }
    }
}
`

I get back this result:

`HTTP/1.1 200 OK
content-encoding: gzip
content-length: 178
content-type: application/json

{
    "data": {
        "environmentNodes": [
            {
                "descendantTargets": [
                    {
                        "name": "service:jmx:rmi:///jndi/rmi://cryostat:10000/jmxrmi",
                        "nodeType": "JVM"
                    },
                    {
                        "name": "service:jmx:rmi:///jndi/rmi://cryostat:9091/jmxrmi",
                        "nodeType": "JVM"
                    },
                    {
                        "name": "service:jmx:rmi:///jndi/rmi://cryostat:9093/jmxrmi",
                        "nodeType": "JVM"
                    },
                    {
                        "name": "service:jmx:rmi:///jndi/rmi://cryostat:9094/jmxrmi",
                        "nodeType": "JVM"
                    },
                    {
                        "name": "service:jmx:rmi:///jndi/rmi://cryostat:9095/jmxrmi",
                        "nodeType": "JVM"
                    },
                    {
                        "name": "service:jmx:rmi:///jndi/rmi://cryostat:9096/jmxrmi",
                        "nodeType": "JVM"
                    }
                ],
                "name": "JDP",
                "nodeType": "REALM"
            }
        ]
    }
}
`
However, if I then do the same query but this time filtering for the `nodeType`:

`query {
    environmentNodes(filter: { name: "JDP", nodeType: "REALM" }) {
        name
        nodeType
        descendantTargets {
            name
            nodeType
        }
    }
}
`

I get back an empty list (i.e. no matches) when I expect to get the same result as the first query:

`HTTP/1.1 200 OK
content-encoding: gzip
content-length: 58
content-type: application/json

{
    "data": {
        "environmentNodes": []
    }
}
`

However, the query works as expected if your filter for `nodeType: "Realm"` instead of `nodeType: "REALM"`, so it looks like a casing mismatch issue on line 78 of the `EnvironmentNodesFetcher`.

andrewazores commented:
Nice catch @hareetd . I had an `equalsIgnoreCase()` on that line previously but it didn't seem right, so I removed it. But that's broken because in the discovery tree result, what the user sees is the `toString()` of the NodeType enum - this is where the uppercase `REALM` comes from. The `getKind()` of that `BaseNodeType.REALM` enum member is `Realm`, though. I updated that condition to accept either form, but I think the proper fix is that the GraphQL queries and the Discovery API should report the nodes' type/kind as the `Realm` form, ie the value specified by the enum members, and not just the enum members' names. I'll see about fixing that separately.

andrewazores commented:
^ https://github.com/cryostatio/cryostat/issues/853

hareetd reviewed (approved):
Looks good.

jan-law reviewed (approved):
lgtm
